### PR TITLE
[Merged by Bors] - feat(Algebra/Lie): existence of Cartan subalgebras

### LINKED ISSUE #10065: Cartan subalgebras exist
We should prove that semisimple Lie algebras have Cartan subalgebras. Furthermore we should prove this making only weak assumptions on the coefficients.

### PR DESCRIPTION
**Pull request by jcommelin**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+* `exists_IsCartanSubalgebra`:
+  A finite-dimensional Lie algebra `L` over an infinite field `K` has a Cartan subalgebra.
+
+## References
+
+* [barnes1967]: "On Cartan subalgebras of Lie algebras" by D.W. Barnes.
+
+-/
+
+-- move this to `Mathlib.Algebra.MvPolynomial.Basic`
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Lie/CartanExists.lean:32]: This seems to have been forgotten.

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+lemma MvPolynomial.coe_aeval_eq_eval {R σ : Type*} [CommRing R] (x : σ → R) :
+    RingHomClass.toRingHom (MvPolynomial.aeval x) = MvPolynomial.eval x := rfl
+
+-- move this to `Mathlib.LinearAlgebra.FiniteDimensional`
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Lie/CartanExists.lean:36]: Same.

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+open FiniteDimensional in
+lemma _root_.Submodule.eq_top_of_finrank
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Lie/CartanExists.lean:38]: Isn't this [Submodule.eq_top_of_finrank_eq](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/FiniteDimensional.html#Submodule.eq_top_of_finrank_eq)?

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+  (LinearMap.polyCharpoly (LieHom.toLinearMap φ) bL).map <| RingHomClass.toRingHom <|
+    MvPolynomial.aeval fun i ↦ C (bL.repr y i) * X + C (bL.repr x i)
+
+lemma lieCharpoly_monic : (lieCharpoly R M x y).Monic :=
+  (polyCharpoly_monic _ _).map _
+
+lemma lieCharpoly_natDegree : (lieCharpoly R M x y).natDegree = finrank R M := by
+  rw [lieCharpoly, (polyCharpoly_monic _ _).natDegree_map, polyCharpoly_natDegree]
+
+lemma lieCharpoly_map_eval (r : R) :
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Lie/CartanExists.lean:119]:
```suggestion
variable {R} in
lemma lieCharpoly_map_eval (r : R) :
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+lemma rank_le_finrank_engel (x : L) :
+    rank K L ≤ finrank K (engel K x) :=
+  (rank_le_natTrailingDegree_charpoly_ad K x).trans
+    (finrank_engel K x).ge
+
+lemma isRegular_iff_finrank_engel_eq_rank (x : L) :
+    IsRegular K x ↔ finrank K (engel K x) = rank K L := by
+  rw [isRegular_iff_natTrailingDegree_charpoly_eq_rank, finrank_engel]
+
+variable {K}
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Lie/CartanExists.lean:73]: I think you can move this line to line 92 with the same effect.

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 3 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:38]: Good catch!

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 5 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:73]: In fact, I could just remove it altogether.

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+    -- We ultimately want to show `engel K x ≤ engel K y`
+    intro z hz
+    -- which holds by definition of Engel subalgebra and the nilpotency that we just established.
+    rw [mem_engel_iff]
+    exact this ⟨z, hz⟩
+  -- To show that `χ = X ^ r`, it suffices to show that all coefficients in degrees `< r` are `0`.
+  suffices ∀ i < r, χ.coeff i = 0 by
+    simp_rw [r, ← lieCharpoly_natDegree K E x u] at this ⊢
+    apply (lieCharpoly_monic K E x u).eq_X_pow_of_natTrailingDegree_eq_natDegree
+    apply le_antisymm (natTrailingDegree_le_natDegree _)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Lie/CartanExists.lean:195]: #12448 can save you one line here :)

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+  -- on a sufficiently large subset `s` of `K`.
+  -- But we first need to get our hands on that subset `s`.
+  -- We start by observing that `ψ` has non-trivial constant coefficient.
+  have hψ : constantCoeff ψ ≠ 0 := by
+    -- Suppose that `ψ` in fact has trivial constant coefficient.
+    intro H
+    -- Then there exists a `z ≠ 0` in `Q` such that `⁅x, z⁆ = 0`.
+    obtain ⟨z, hz0, hxz⟩ : ∃ z : Q, z ≠ 0 ∧ ⁅x, z⁆ = 0 := by
+      -- Indeed, if the constant coefficient of `ψ` is trivial,
+      -- then `0` is a root of -- the characteristic polynomial of `⁅0 • u + x, _⁆` acting on `Q`,
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Lie/CartanExists.lean:242]:
```suggestion
      -- then `0` is a root of the characteristic polynomial of `⁅0 • u + x, _⁆` acting on `Q`,
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+  use ((LieModule.toEndomorphism K U Q) y ^ k) z'
+  refine ⟨?_, ?_⟩
+  · -- And `⁅y, _⁆ ^ k` applied to `z'` is non-zero by definition of `n`.
+    apply Nat.find_min hz'; omega
+  · rw [← hn, hk, pow_succ', LinearMap.mul_apply]
+
+variable (K L)
+
+open Cardinal in
+lemma exists_IsCartanSubalgebra_of_finrank_le_card (h : finrank K L ≤ #K) :
```
Comment (Ruben-VandeVelde) [inline Mathlib/Algebra/Lie/CartanExists.lean:359]: I think `exists_isCartanSubalgebra_of_finrank_le_card` with lower-case `i`, also in the next lemma

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 6 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:195]: Nice!

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 8 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Thanks, fixed

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 6 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Fixed

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+local notation "φ" => LieModule.toEndomorphism R L M
+
+/-- Let `x` and `y` be elements of a Lie `R`-algebra `L`, and `M` a Lie module over `M`.
+Then the characteristic polynomials of the family of endomorphisms `⁅r • y + x, _⁆` of `M`
+have coefficients that are polynomial in `r : R`.
+In other words, we obtain a polynomial over `R[X]`
+that specializes to the characteristic polynomial of `⁅r • y + x, _⁆` under the map `X ↦ r`.
+This polynomial is captured in `lieCharpoly R M x y`. -/
+private noncomputable
+def lieCharpoly : Polynomial R[X] :=
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:93]:
```suggestion
def lieCharpoly : R[X][X] :=
```
Or maybe even better if we used the [PolynomialPolynomial scope](https://github.com/leanprover-community/mathlib4/blob/8d0c6e6f821392660468fdce7c5463740d7988be/Mathlib/AlgebraicGeometry/EllipticCurve/Affine.lean#L84) then this could be `R[X][Y]`. (This would require moving that scope to a more basic file.)

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+Given a Lie subalgebra `U` of `L`, and an element `x ∈ U` such that `U ≤ engel K x`.
+Suppose that `engel K x` is minimal amongst the Engel subalgebras `engel K y` for `y ∈ U`.
+Then `engel K x ≤ engel K y` for all `y ∈ U`.
+
+Lemma 2 in [barnes1967]. -/
+lemma engel_le_engel (hLK : finrank K L ≤ #K)
+    (U : LieSubalgebra K L) (x : U) (hUx : U ≤ engel K (x : L))
+    (hmin : ∀ y : U, engel K (y : L) ≤ engel K (x : L) → engel K (y : L) = engel K (x : L))
+    (y : U) :
+    engel K (x : L) ≤ engel K (y : L) := by
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:150]: How about introducing the following somewhere in the order theory basics:
```lean
def IsMinimumOfMinimal {ι α : Type*} [Preorder α] (f : ι → α) : Prop :=
    ∀ i, (∀ j, f j ≤ f i → f j = f i) → ∀ j, f i ≤ f j
```
and then restating this as:
```suggestion
lemma isMinimumOfMinimal_engel_of_finrank_le (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L) :
    IsMinimumOfMinimal <| engel K ∘ U.incl := by
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+end CommRing
+
+section Field
+
+variable {K L : Type*}
+variable [Field K] [LieRing L] [LieAlgebra K L] [Module.Finite K L]
+
+open FiniteDimensional LieSubalgebra Module.Free Polynomial
+
+open Cardinal LieModule LieSubalgebra Polynomial engel_le_engel in
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:139]:
```suggestion
open FiniteDimensional LieSubalgebra Polynomial Cardinal LieModule engel_le_engel

```
and then you can drop the `open Cardinal in` just above `exists_isCartanSubalgebra_of_finrank_le_card`

**Review by ocfnash — COMMENTED** [review]
I'll review this properly later today when I have more time. For now, just some superficial remarks.

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+  rw [← mul_one j, lieCharpoly, coeff_map]
+  apply MvPolynomial.aeval_natDegree_le
+  · apply (polyCharpoly_coeff_isHomogeneous φ (chooseBasis R L) _ _ hij).totalDegree_le
+  intro k
+  apply Polynomial.natDegree_add_le_of_degree_le
+  · apply (Polynomial.natDegree_C_mul_le _ _).trans
+    simp only [natDegree_X, le_rfl]
+  · simp only [natDegree_C, zero_le]
+
+end engel_le_engel
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:128]: I think this entire section (from `lieCharpoly` to `lieCharpoly_coeff_natDegree`) should be done for a general family of linear maps `L →ₗ[R] Module.End R M` and moved to the file `Mathlib.Algebra.Module.LinearMap.Polynomial`.

Given a pair of endomorphisms $f$ and $g$, I think it's quite plausible that someone else will want results about the characteristic polynomial of a family $s ↦ sf + g$ at some point.

With this in mind I think it is worth the small additional effort to make the definition depend only on the two endomorphisms `f` and `g` and then add some glue to join up with the case that we have a full linear family `φ : L →ₗ[R] Module.End R M` with `f = φ x` and `g = φ y` for some `x` and `y`.

So the basic definition would be (hopefully this can be cleaned up):
```lean
-- Or `LinearMap.charpoly₂`
noncomputable def LinearMap.charpolyOfPair [Module K M] [Module.Finite K M]
    (f g : Module.End K M) : K[X][X] :=
  letI bL := chooseBasis K (Submodule.span K {f, g})
  (LinearMap.polyCharpoly (Submodule.subtype _) bL).map <| RingHomClass.toRingHom <|
    MvPolynomial.aeval fun i ↦ C (bL.repr ⟨g, Submodule.subset_span <| by simp⟩ i) * X +
      C (bL.repr ⟨f, Submodule.subset_span <| by simp⟩ i)
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+    · omega
+    · dsimp only [r] at hi ⊢
+      rw [Nat.add_sub_cancel' hi.le]
+  -- We need to show that for all `α ∈ s`, the polynomial `coeff χ i` evaluates to zero at `α`.
+  intro α hα
+  -- Once again, we are left with showing that `⁅y, _⁆` acts nilpotently on `E`.
+  rw [← coe_evalRingHom, ← coeff_map, lieCharpoly_map_eval,
+    (LinearMap.charpoly_eq_X_pow_iff _).mpr, coeff_X_pow, if_neg hi.ne]
+  -- To do so, it suffices to show that the Engel subalgebra of `y = a • u + x` is contained in `E`.
+  let y := α • u + x
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:305]: I was briefly confused by the fact that this shadows the existing `y`:
```suggestion
  -- To do so, it suffices to show that the Engel subalgebra of `v = a • u + x` is contained in `E`.
  let v := α • u + x
```
followed by a bunch of trivial fixes below ofc

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+    -- Then there exists a `z ≠ 0` in `Q` such that `⁅x, z⁆ = 0`.
+    obtain ⟨z, hz0, hxz⟩ : ∃ z : Q, z ≠ 0 ∧ ⁅x, z⁆ = 0 := by
+      -- Indeed, if the constant coefficient of `ψ` is trivial,
+      -- then `0` is a root of the characteristic polynomial of `⁅0 • u + x, _⁆` acting on `Q`,
+      -- and hence we find an eigenvector `z` as desired.
+      apply_fun (evalRingHom 0) at H
+      rw [constantCoeff_apply, ← coeff_map, lieCharpoly_map_eval,
+        ← constantCoeff_apply, map_zero, LinearMap.charpoly_constantCoeff_eq_zero_iff] at H
+      simpa only [coe_bracket_of_module, ne_eq, zero_smul, zero_add,
+        LieModule.toEndomorphism_apply_apply] using H
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:247]:
```suggestion
        toEndomorphism_apply_apply] using H
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+  -- Now we are in good shape.
+  -- Fix an element `z` in the Engel subalgebra of `y`.
+  intro z hz
+  -- We need to show that `z` is in `E`, or alternatively that `z = 0` in `Q`.
+  show z ∈ E
+  rw [← LieSubmodule.Quotient.mk_eq_zero]
+  -- We denote the image of `z` in `Q` by `z'`.
+  set z' : Q := LieSubmodule.Quotient.mk' E z
+  -- First we observe that `z'` is killed by a power of `⁅y, _⁆`.
+  have hz' : ∃ n : ℕ, ((toEndomorphism K U Q) y ^ n) z' = 0 := by
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:322]:
```suggestion
  have hz' : ∃ n : ℕ, ((toEndomorphism K U Q y) ^ n) z' = 0 := by
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+    rw [LieModuleHom.map_zero] at hn
+    rw [← hn]
+    clear hn
+    induction n with
+    | zero => simp only [Nat.zero_eq, pow_zero, LinearMap.one_apply]
+    | succ n ih => rw [pow_succ', pow_succ', LinearMap.mul_apply, ih]; rfl
+  classical
+  -- Now let `n` be the smallest power such that `⁅y, _⁆ ^ n` kills `z'`.
+  set n := Nat.find hz' with _hn
+  have hn : ((toEndomorphism K (↥U) Q) y ^ n) z' = 0 := Nat.find_spec hz'
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:336]:
```suggestion
  have hn : ((toEndomorphism K U Q y) ^ n) z' = 0 := Nat.find_spec hz'
```

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 17]
```diff
@@ -0,0 +1,380 @@
...
+  -- If `n = k + 1`, then we can write `⁅y, _⁆ ^ n = ⁅y, _⁆ ∘ ⁅y, _⁆ ^ k`.
+  -- Recall that `constantCoeff ψ` is non-zero on `α`, and `y = α • u + x`.
+  specialize hsψ α hα
+  -- Hence `⁅y, _⁆` acts injectively on `Q`.
+  rw [← coe_evalRingHom, constantCoeff_apply, ← coeff_map, lieCharpoly_map_eval,
+    ← constantCoeff_apply, ne_eq, LinearMap.charpoly_constantCoeff_eq_zero_iff] at hsψ
+  -- We deduce from this that `z' = 0`, arguing by contraposition.
+  contrapose! hsψ
+  -- Indeed `⁅y, _⁆` kills `⁅y, _⁆ ^ k` applied to `z'`.
+  use ((LieModule.toEndomorphism K U Q) y ^ k) z'
```
Comment (ocfnash) [inline Mathlib/Algebra/Lie/CartanExists.lean:349]:
```suggestion
  use ((toEndomorphism K U Q y) ^ k) z'
```

**Review by ocfnash — COMMENTED** [review]
I finally looked at the giant proof. I think it's fine as is for the following reasons:

- It's very well commented, and quite easy to follow as a result.
- It's long but not ludicrous, coming in at c.125 lines without comments.
- I could not find any useful results to split out (this would be surprising except that I know you have already made this effort).
- I encounter no speed issues with it, at least on my (admittedly fast) laptop. The profiler reports `linting took 308ms; elaboration took 171ms`.

I do however note Patrick's good point that notation is no reason to avoid splitting it.

In summary: perhaps with even greater effort there might be a useful split here but I see none and I see no cost to keeping this as it stands.

File: Mathlib/Algebra/Lie/CartanExists.lean [hunk 18]
```diff
@@ -0,0 +1,380 @@
...
+  (ad K L x).finrank_maximalGeneralizedEigenspace
+
+lemma rank_le_finrank_engel (x : L) :
+    rank K L ≤ finrank K (engel K x) :=
+  (rank_le_natTrailingDegree_charpoly_ad K x).trans
+    (finrank_engel K x).ge
+
+lemma isRegular_iff_finrank_engel_eq_rank (x : L) :
+    IsRegular K x ↔ finrank K (engel K x) = rank K L := by
+  rw [isRegular_iff_natTrailingDegree_charpoly_eq_rank, finrank_engel]
```
[the remark this answers is not in the document]
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:59]: That would imply adding
```lean
import Mathlib.Algebra.Lie.EngelSubalgebra
import Mathlib.LinearAlgebra.Eigenspace.Zero
```
to that file. Is that ok?

These lemmas bring together a bunch of developments... Should I start a small file `RankLemmas.lean`?

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 13 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Thanks! Good suggestion.

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 10 above)
Comment (ocfnash) [reply Mathlib/Algebra/Lie/CartanExists.lean:149]: I think it does. For example if you start like this then the proof works with no modification:
```
lemma isMinimumOfMinimalOn_engel_of_finrank_le
    (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L) :
    IsMinimumOfMinimalOn {x : U | U ≤ engel K (x : L)} (engel K ∘ U.incl) := by
  intro x (hUx : U ≤ engel K (x : L)) hmin y
  change ∀ y : U, engel K (y : L) ≤ engel K (x : L) → engel K (y : L) = engel K (x : L) at hmin
  change engel K (x : L) ≤ engel K (y : L)
```

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 18 above)
[the remark this answers is not in the document]
Comment (ocfnash) [reply Mathlib/Algebra/Lie/CartanExists.lean:59]: I'd add the imports as it's hard to imagine any consumer of `Mathlib.Algebra.Lie.Rank` who won't quickly combine it with these files but if you're not convinced then feel free to either leave these lemmas where they are or start a new file.

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 12 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: I think that I would first want to see more evidence that this is useful in other places, if you don't mind.

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 10 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:148]: Weird, I don't know what I was trying this afternoon. But you are absolutely right!

Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean:148]: I've been playing a bit more with the statement. What do you think of the following version:
```lean
lemma engel_le_engel (hLK : finrank K L ≤ #K) (U : LieSubalgebra K L)
    (E : {engel K x | x ∈ U}) (hUE : U ≤ E) (hE : IsMin E) :
    IsBot E := by
```
To me it seems even closer to the informal statement.

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 18 above)
[the remark this answers is not in the document]
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Done

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 10 above)
Comment (jcommelin) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Done

File: Mathlib/Algebra/Lie/CartanExists.lean (hunk 12 above)
Comment (ocfnash) [reply Mathlib/Algebra/Lie/CartanExists.lean]: Agreed; when I wrote that I did try to think of examples and the best I could come up with were examples of situations where one cares about $\det(sf + g)$ or $\tr(sf + g)$ but I admit I couldn't think about cases (except here!) where one needed the full characteristic polynomial. I think it's totally reasonable to leave this as it stands.

**Comment by ocfnash** [comment]
Thanks!

bors merge
